### PR TITLE
statistics: separate Column and Index Histogram

### PR DESCRIPTION
**Pull request by coocood**
Column and Index use histogram in different way, we should separate them for more
accurate estimation.

### CODE REVIEW

**Comment by coocood** [comment]
@hanfei1991 @lamxTyler PTAL

File: statistics/builder.go [hunk 1]
```diff
@@ -48,19 +48,19 @@ type buildStatsTask struct {
 	err   error
 	index bool
 	t     *Table
```
Comment (hanfei1991) [inline statistics/builder.go]: t is still useful?

File: statistics/histogram.go [hunk 2]
```diff
@@ -100,125 +100,129 @@ func colStatsFromStorage(ctx context.Context, tableID int64, colID int64, tp *ty
...
 }
 
-func (c *Column) String() string {
-	strs := make([]string, 0, len(c.Buckets)+1)
-	strs = append(strs, fmt.Sprintf("column:%d ndv:%d", c.ID, c.NDV))
-	for _, bucket := range c.Buckets {
+func (hg *Histogram) toString(isIndex bool) string {
+	strs := make([]string, 0, len(hg.Buckets)+1)
+	if isIndex {
+		strs = append(strs, fmt.Sprintf("column:%d ndv:%d", hg.ID, hg.NDV))
```
Comment (lamxTyler) [inline statistics/histogram.go]: This should be `index:...`?

File: statistics/builder.go [hunk 3]
```diff
@@ -274,34 +288,36 @@ func (b *Builder) buildColumn(t *Table, sc *variable.StatementContext, id int64,
...
 			})
 		}
 	}
-	return col, nil
+	return hg, nil
 }
 
-func copyFromIndexColumns(ind *Column, id int64) (*Column, error) {
-	col := &Column{
+// index histogram is encoded, it need to be decoded to be used as column histogram.
```
Comment (lamxTyler) [inline statistics/builder.go]: Does the `index` need capitalize?

File: statistics/histogram.go [hunk 4]
```diff
@@ -232,20 +236,39 @@ func (c *Column) lowerBound(sc *variable.StatementContext, target types.Datum) (
...
+}
+
+func (c *Column) String() string {
+	return c.Histogram.toString(false)
+}
+
+// Index represents an Index histogram.
+type Index struct {
+	Histogram
+	NumColumns int
```
Comment (lamxTyler) [inline statistics/histogram.go]: I noticed that this value is setted somewhere, but not used. Does it have some future usage?

File: statistics/builder.go [hunk 5]
```diff
@@ -48,19 +48,19 @@ type buildStatsTask struct {
...
-func (b *Builder) buildMultiColumns(t *Table, IDs []int64, baseOffset int, isSorted bool) {
+func (b *Builder) buildMultiHistograms(t *Table, IDs []int64, baseOffset int, isSorted bool) {
 	for i, id := range IDs {
 		var err error
-		var col *Column
+		var hg *Histogram
 		if isSorted {
-			col, err = b.build4SortedColumn(t, b.Ctx.GetSessionVars().StmtCtx, id, b.IdxRecords[i+baseOffset], b.NumBuckets, false)
-			b.doneCh <- &buildStatsTask{err: err, index: true, t: t, col: col}
+			hg, err = b.build4SortedColumn(t, b.Ctx.GetSessionVars().StmtCtx, id, b.IdxRecords[i+baseOffset], b.NumBuckets, false)
```
Comment (lamxTyler) [inline statistics/builder.go]: Why not `hg, err :=...`?

File: statistics/builder.go [hunk 6]
```diff
@@ -130,45 +133,56 @@ func (b *Builder) NewTable() (*Table, error) {
...
-		if len(idx.Columns) == 1 {
-			for _, col := range b.TblInfo.Columns {
-				if col.Name.L == idx.Columns[0].Name.L && t.Columns[col.ID] == nil {
-					t.Columns[col.ID], err = copyFromIndexColumns(t.Indices[idx.ID], col.ID)
-					if err != nil {
-						return nil, errors.Trace(err)
-					}
-					break
-				}
+	// The sorted bucket is more accurate, we replace the sampled column histogram with index histogram if the
```
Comment (lamxTyler) [inline statistics/builder.go]: `sorted bucket`?

File: statistics/builder.go [hunk 7]
```diff
@@ -130,45 +133,56 @@ func (b *Builder) NewTable() (*Table, error) {
...
+func indexNumColumnsByID(tblInfo *model.TableInfo, idxID int64) int {
+	for _, idx := range tblInfo.Indices {
+		if idx.ID == idxID {
+			return len(idx.Columns)
+		}
+	}
+	return 0
+}
+
 // build4SortedColumn builds column statistics for sorted columns.
```
Comment (lamxTyler) [inline statistics/builder.go]: `builds histogram...` is better?

File: statistics/histogram.go [hunk 8]
```diff
@@ -232,20 +236,39 @@ func (c *Column) lowerBound(sc *variable.StatementContext, target types.Datum) (
...
+// Column represents a column histogram.
+type Column struct {
+	Histogram
+}
+
+func (c *Column) String() string {
+	return c.Histogram.toString(false)
+}
+
+// Index represents an Index histogram.
```
Comment (lamxTyler) [inline statistics/histogram.go]: Why capitalize the second `Index`?

**Comment by lamxTyler** [comment]
A little question for line [187](https://github.com/pingcap/tidb/blob/master/statistics/table.go#L187). 
Here the `Columns[...]` is type of `Column`,yet `BetweenRowCount` is a method of `Histgram`, is it a good practise?

File: statistics/histogram.go (hunk 4 above)
Comment (coocood) [inline statistics/histogram.go]: Yes, the code in `plan/statistics_helper.go` will be moved here, `NumColumns` will be used then.

File: statistics/builder.go (hunk 6 above)
Comment (coocood) [inline statistics/builder.go]: I think `the` is needed.

**Comment by coocood** [comment]
@lamxTyler @hanfei1991 PTAL

File: statistics/builder.go (hunk 6 above)
Comment (lamxTyler) [inline statistics/builder.go]: I mean, it is not the bucket sorted. It is the original data/column sorted.

**Comment by shenli** [comment]
What is the relationship between `Column`, `Index` and `Histgram`?

File: statistics/builder.go [hunk 9]
```diff
@@ -202,54 +213,54 @@ func (b *Builder) build4SortedColumn(t *Table, sc *variable.StatementContext, id
...
 	}
 	atomic.StoreInt64(&t.Count, count)
-	return col, nil
+	return hg, nil
 }
 
-// buildColumn builds column statistics from samples.
-func (b *Builder) buildColumn(t *Table, sc *variable.StatementContext, id int64, ndv int64, samples []types.Datum, bucketCount int64) (*Column, error) {
+// buildColumn builds histogram from samples for column.
+func (b *Builder) buildColumn(t *Table, sc *variable.StatementContext, id int64, ndv int64, samples []types.Datum, bucketCount int64) (*Histogram, error) {
```
Comment (shenli) [inline statistics/builder.go]: So we should rename it to buildHistogram?

Comment (coocood) [inline statistics/builder.go]: No, it is specifically for build column.

**Comment by coocood** [comment]
@shenli 
In OOP, `Column` and `Index` are sub class of `Histogram`.

**Comment by shenli** [comment]
LGTM

**Pull request merged**